set a user agent header in the requests sent by buildsys

vyaghras opened the pull request:
Issue number:
 Resolves [2903](https://github.com/bottlerocket-os/bottlerocket/issues/2903) 

Description of changes:
```
 Buildsys (via reqwest) does not seem to set a user agent by default, and this causes netfilter.org to 
 respond 403 Forbidden and the build to fail if the package is unavailable in the look-aside cache. 
 By this change we will add USER_AGENT as a header in HTTP request. 
```
Testing done:
All unit tests passes.

Terms of contribution:

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

markusboehme commented:
The change itself looks good to me. Nice! :champagne:

I assume you have tested the change actually fixes #2903. If so, please include the rough sketch of steps you took (or full commands if they're more involved) for any reviewer who might want to follow along.

Some comments on the form that would be good to get straightened out, because they'll be helpful for all further changes as well:

* The commit does not have a proper author and committer recorded ("EC2 Default User"). Please set one (`git config --global user.name 'Your Name'`, `git config --global user.email 'youremail@example.com'`) and recreate the commit.
* Try to include more of the reasoning for your change in your commit message. The description of your changes in this PR would be helpful to have in the actual commit message because it actually lays out *why* you're making the change you do. By including it in the actual commit message, the Git history stands on its own without having to refer to GitHub PRs.
* Projects often aim to follow a certain style for their commit messages so at a later point it's easier to quickly scan them for anything that might be relevant to some research/debugging one is doing. I often take a look at commits in a similar area to the one I'm committing to using e.g. `git log -- tools/buildsys` so I can try to adhere to that style when I'm writing my own commit messages. Here, a `buildsys: ` prefix on the first line of the commit message would signify what area you're modifying. Also, the more complete description of the change should be wrapped at 72 characters.

vyaghras commented:
This change fixes [2903](https://github.com/bottlerocket-os/bottlerocket/issues/2903) . Following steps has been followed to verify: 
1. Build failed with libftnl latest version [libnftnl-1.2.5.tar.xz] without this change using command `cargo make -e BUILDSYS_UPSTREAM_SOURCE_FALLBACK="true"` (to make sure we are fetching from upstream).
2. Build passes with libftnl latest version [libnftnl-1.2.5.tar.xz] with this change using command `cargo make -e BUILDSYS_UPSTREAM_SOURCE_FALLBACK="true"`and it downloads libnftnl-1.2.5.tar.xz.

vyaghras committed:
buildsys: Add user agent header in http request

Buildsys does not set the user agent in HTTP request when fetching
upstream sources. But some servers may not like the requests without a
user agent header and netfilter.org is one of them. All these
requests succeed when a user agent is set, so include the header in
requests that buildsys sends out.

yeazelm commented on tools/buildsys/src/cache.rs:
```diff
@@ -77,8 +78,20 @@ impl LookasideCache {
+        headers.insert(
+            USER_AGENT,
+            HeaderValue::from_static(
+                "Bottlerocket buildsys (https://github.com/bottlerocket-os/bottlerocket)",
```
Is there an easy way to have the version from the repo included? It's typical to have a version in the UserAgent.

stmcginnis replied:
It would be a little extra work, but you could use a `std::process::Command` to call out and get `git rev-parse --short HEAD` for the repo, and/or read in the `Release.toml` file to get the version string.

markusboehme replied:
If we do want this, we could read one of the environment variables predefined by `Makefile.toml`. The most appropriate for this use to me seems `BUILDSYS_VERSION_IMAGE`, expanding to e.g. `1.13.0`.

vyaghras replied:
I too feel  that BUILDSYS_VERSION_IMAGE or BUILDSYS_VERSION_FULL could be good option to pass here.

stmcginnis replied:
Good call!

yeazelm reviewed (commented):
I'd like to see the before/after output of those tests you ran in the comments of the PR just so we can see the actual change.

vyaghras commented:
Befor this change:
```   Compiling libnftnl v0.1.0 (/home/ec2-user/bottlerocket/packages/libnftnl)
   Compiling libelf v0.1.0 (/home/ec2-user/bottlerocket/packages/libelf)
   Compiling containerd v0.1.0 (/home/ec2-user/bottlerocket/packages/containerd)
   Compiling os v0.1.0 (/home/ec2-user/bottlerocket/packages/os)
   Compiling grep v0.1.0 (/home/ec2-user/bottlerocket/packages/grep)
   Compiling ecr-credential-provider v0.1.0 (/home/ec2-user/bottlerocket/packages/ecr-credential-provider)
error: failed to run custom build command for `libnftnl v0.1.0 (/home/ec2-user/bottlerocket/packages/libnftnl)`

Caused by:
  process didn't exit successfully: `/home/ec2-user/bottlerocket/variants/target/x86_64/debug/build/libnftnl-3b082d0389bbda22/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-changed=Cargo.toml
  Fetching "libnftnl-1.2.5.tar.xz" from upstream source

  --- stderr
  Failed to fetch 'https://cache.bottlerocket.aws/libnftnl-1.2.5.tar.xz/576ccd0815063a6ef3095b5514c3d286b4450fad98fbf7a85cd537f66adf043e7e6295d4c84cc3cbfd18cf9a29576d15a88cc439a61a3e654841a27c71babea1/libnftnl-1.2.5.tar.xz': 403 Forbidden
  ExternalFileFetch: Failed to fetch 'http://netfilter.org/projects/libnftnl/files/libnftnl-1.2.5.tar.xz': 403 Forbidden
warning: build failed, waiting for other jobs to finish...
[cargo-make] ERROR - Error while executing command, exit code: 101
[cargo-make] WARN - Build Failed.
```
After this change:
```
[cargo-make] INFO - Running Task: build-variant
   Compiling libnftnl v0.1.0 (/home/ec2-user/bottlerocket/packages/libnftnl)
   Compiling libelf v0.1.0 (/home/ec2-user/bottlerocket/packages/libelf)
   Compiling containerd v0.1.0 (/home/ec2-user/bottlerocket/packages/containerd)
   Compiling grep v0.1.0 (/home/ec2-user/bottlerocket/packages/grep)
   Compiling ecr-credential-provider v0.1.0 (/home/ec2-user/bottlerocket/packages/ecr-credential-provider)
   Compiling os v0.1.0 (/home/ec2-user/bottlerocket/packages/os)
   Compiling kubernetes-1_24 v0.1.0 (/home/ec2-user/bottlerocket/packages/kubernetes-1.24)
   Compiling makedumpfile v0.1.0 (/home/ec2-user/bottlerocket/packages/makedumpfile)
   Compiling iptables v0.1.0 (/home/ec2-user/bottlerocket/packages/iptables)
   Compiling release v0.0.0 (/home/ec2-user/bottlerocket/packages/release)
   Compiling aws-k8s-1_24 v0.1.0 (/home/ec2-user/bottlerocket/variants/aws-k8s-1.24)
    Finished dev [optimized] target(s) in 1m 20s
[cargo-make] INFO - Build Done in 92.07 seconds.
```

stmcginnis reviewed (approved):
It makes me a little apprehensive that we now have an implicit requirement that the `BUILDSYS_VERSION_FULL` variable is set. It might be nice to have `getenv` fallback to some sort of reasonable default value rather than erroring out.

That said, I don't think it's really a big deal, and at least so far we don't have any use cases for running this outside of the `Makefile.toml` tasks that will have this set. So just noting it here, but I don't feel strongly that we need to change anything.

webern commented:
> implicit requirement that the BUILDSYS_VERSION_FULL variable is set

There are a lot of these implicit env var requirements. We should probably surface these all in clap::Args structs.
